Source/Core/Core/HW/SI/SI.cpp

```diff
@@ -416,10 +417,17 @@ void RegisterMMIO(MMIO::Mapping* mmio, u32 base)
   // Register SI buffer direct accesses.
   for (size_t i = 0; i < s_si_buffer.size(); i += sizeof(u32))
   {
-    const u32 address = base | static_cast<u32>(s_si_buffer.size() + i);
+    const u32 address = base | static_cast<u32>(SI_IO_BUFFER + i);
 
-    mmio->Register(address, MMIO::DirectRead<u32>((u32*)&s_si_buffer[i]),
-                   MMIO::DirectWrite<u32>((u32*)&s_si_buffer[i]));
+    mmio->Register(address, MMIO::DirectSwapRead<u32>((u32*)&s_si_buffer[i]),
+                   MMIO::DirectSwapWrite<u32>((u32*)&s_si_buffer[i]));
```

Comment:
Are those casts actually ok, or do they fall within undefined behavior land?

Reply:
They're undefined behavior (as are the C-style ones that cast away volatile), I didn't call them out in my review since they were pre-existing (easiest way to get rid of the need for volatile ones is to get rid of the use of volatile within [SCPFifoStruct](https://github.com/dolphin-emu/dolphin/blob/0bf24f549a2cc6146e138bd03dd1338689d96b3f/Source/Core/VideoCommon/CommandProcessor.h#L17) and it's related [MMIO](https://github.com/dolphin-emu/dolphin/blob/9f4d5122a2a1b2d0dbfd6164a616d1c859ef0d0b/Source/Core/VideoCommon/CommandProcessor.cpp#L150) code, but that's definitely out of scope).